### PR TITLE
Make conversions from/to sparse semi-structured always @torch.compile-d

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105272

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Comment by alexsamardzic** [comment]
Internal conversion functions from/to sparse semi-structured representation are written and optimized so that they could be `@torch.compile`-d.  This PR is just to discuss would it be possible/advisable to have `@torch.compile`-d functions in the PyTorch source tree.  As far as I can see, there is only one function implemented this way, it's `hash_storage()`, in `torch/utils/_content_store.py`.  It has some guards and checks to verify that the compilation is possible, so maybe these could be extracted in a helper (as suggested by a comment there) through this PR too.

One of issues here is what happens when `@torch.compile`-d functions is called from another function or module that is also compiled.  For simple case like:
```Python
import torch

@torch.compile
def myadd(x, y):
    return x + y

@torch.compile
def myaddplus2(x, y):
    return myadd(x, y) + 2

x = torch.ones((5,), dtype=torch.float, device="cuda")
y = torch.ones((5,), dtype=torch.float, device="cuda")
print(myaddplus2(x, y))
```
it turns out that the single Triton kernel is generated, but it is unsure will this always be the case.

(Another question arose throughout the development of these functions, and it was about possibility of setting `torch._inductor.config.aggressive_fusion` temporarily, i.e. only for the compilation of specific functions.  However, it seems at the moment that the same Triton code is generated for mentioned conversion functions with or without having it set, so this question is not relevant here.)

cc @lezcano @peterbell10

**Comment by lezcano** [comment]
cc @malfet @ezyang

**Comment by ezyang** [comment]
I wrote up an experience report about forcing torch.compile on storage_hash. See https://github.com/pytorch/pytorch/issues/101117

It's not strictly necessary to replicate the is supported utilities for your case; in my case, I needed a new primitive that I didn't want to pay eager compiled binary cost for, so the decomposed code literally does not work in eager mode. (For your case, torch.compile should transparently fall back to eager.)

Some of the other problems are still valid. Fortunately, I'm guessing sparse semi-structured is uncommon enough that we should be able to keep on top of the issues. LAND AWAYYYY

**Review by ezyang — APPROVED** [review]

**Comment by lezcano** [comment]
The failure is real.

File: torch/sparse/_semi_structured_conversions.py [hunk 1]
```diff
@@ -1,7 +1,22 @@
...
+    compile_supported = torch._dynamo.is_dynamo_supported()
+    if device_type == "cpu":
+        pass
+    elif device_type == "cuda" and compile_supported:
+        from torch._inductor.utils import has_triton
+
+        compile_supported = has_triton()
+    else:
+        compile_supported = False
+    return compile_supported
```
Comment (ezyang) [inline torch/sparse/_semi_structured_conversions.py:16]: Nope, find a good spot for this (maybe torch._dynamo.utils) and factor both uses of it to use this one.

**Review by ezyang — CHANGES_REQUESTED** [review]

File: torch/sparse/_semi_structured_conversions.py (hunk 1 above)
Comment (alexsamardzic) [reply torch/sparse/_semi_structured_conversions.py:16]: Sure, that was my intention - just wanted to see is it going to pass this failing test this way.

Comment (alexsamardzic) [reply torch/sparse/_semi_structured_conversions.py]: Fixed now.

**Comment by cpuhrsch** [comment]
@alexsamardzic - Please request re-review from blocking reviewers once CI runs green.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**